### ISSUE TITLE
Storybook 4 not compatible with react 15

### ISSUE THREAD

**Issue by Sujimoshi** [storybooks/storybook#4191]
### Bug
Expected: Min version of react is set to `react >= 15.0.0` , but real minimal version of `react` set by `@emotion/core` which need `react >= 16.3.0`
Using react 15 produce error on loading storybook
![image](https://user-images.githubusercontent.com/10153611/45650818-30d88a80-bad8-11e8-985d-b0ce61a745ad.png)

### Please specify which version of Storybook and optionally any affected addons that you're running

- @storybook/react 4.0.0-alpha.21
- react 15

### Screenshots / Screencast / Code Snippets (Optional)
```js
// @storybook/core/package.json
"dependencies": {
    "@emotion/core": "0.13.0",
},
"peerDependencies": {
    "react": ">=15.0.0",
    "react-dom": ">=15.0.0"
},
```
```js
// @emotion/core/package.json
"peerDependencies": {
    "react": ">=16.3.0"
},
```

**Comment by Sujimoshi** [comment]
It's frustrating, since I can not use the react 15 and webpack 4 in my application if it use storybook

**Comment by ndelangen** [comment]
Hey @Sujimoshi thanks for the issue.

We want to help you as much as wel can. I hope you understand we're doing the best we can to stay backwards compatible AND allow users to use the latest and greatest AND have a nice experience doing it ourselves.

Right now this is these thoughts on the matter:
We're moving forward with React 16 compatibility only for the moment. We didn't want to exclude React 15, it just happened with us ourselves updating to the modern tools/libs.

This may sounds strange, but React is the only lib that has this problem because of how we're using webpack & set up the package dependencies. We want to change this.

Mid term, we want to split the manager & preview more and this would allow you to run any version of react, and us to run our own.

Short term there are a couple of options:
- Keep using the alpha.20
- Upgrade to React 16
- Find some compatibility lib
- Assist us with achieving the above faster. I'd be super happy to have a chat about it in detail.
- Assist us in bringing back compatibility using some other method

I understand this is not exactly the news you were hoping to get, Let's help each other!

**Comment by Sujimoshi** [comment]
Isn't alpha.20 have the same dependencies?

**Comment by Sujimoshi** [comment]
Isn't alpha.20 have the same dependencies?

**Comment by Sujimoshi** [comment]
I am not familiar with storybook source code, but why not just remove react from peerDeps, and allow storybook webpack bundle two versions of react

**Comment by brochington** [comment]
I am in the process of updating a React 15 application to use Webpack 4 and Babel 7. An upgrade to React 16 is in the horizon, but our app is quite large, and will take quite a long time.

Of the options that @ndelangen posted above, I chose to pin the version of Storybook. 

Just a note that I needed to pin to version 4.0.0-alpha.18 to get things to work.

**Comment by Hypnosphi** [comment]
@Sujimoshi how does your `.storybook/webpack.config.js` look? There's a chance we could find out how to make it work with `storybook@3.4`

**Comment by Sujimoshi** [comment]
> Just a note that I needed to pin to version 4.0.0-alpha.18 to get things to work.
Thank for you reply, my project is huge, so i can't just update to react 16. But pin to 4.0.0-alpha.18 make sense.

**Comment by Sujimoshi** [comment]
> @Sujimoshi how does your `.storybook/webpack.config.js` look? There's a chance we could find out how to make it work with `storybook@3.4`

I'm in no hurry to switch to webpack 4. I just would like to solve the problem with backward compatibility, and help Storybook. Anyway here is my webpack 4 config: 
```js
const path = require('path')
const OneToOnePlugin = require('webpack-one-to-one-plugin')

module.exports = (unused, args) => {
  const env = args.mode || 'production'
  const isProd = env === 'production'

  const config = {}
  config.module = { rules: [] }
  config.plugins = []

  config.mode = isProd ? 'production' : 'development'

  config.entry = path.resolve(__dirname, 'src/index.js')

  // config.devtool = 'cheap-module-source-map'

  config.context = path.resolve(__dirname, 'src')

  config.output = {
    path: path.resolve(__dirname, 'dist'),
    filename: 'index.js',
    library: 'cond-component-datepickers',
    libraryTarget: 'umd',
  }

  if (isProd) config.plugins.push(new OneToOnePlugin())

  config.module.rules.push({
    test: /\.js$/,
    exclude: /(node_modules|bower_components)/,
    use: {
      loader: 'babel-loader',
    },
  })

  config.module.rules.push({
    test: /\.css$/,
    use: [
      { loader: 'style-loader' },
      { loader: 'css-loader' },
    ],
  })

  config.module.rules.push({
    test: /\.scss$/,
    use: [
      { loader: 'style-loader' },
      { loader: 'css-loader' },
      { loader: 'sass-loader' }],
  })

  config.module.rules.push({
    test: /\.(png|svg|jpg|gif)$/,
    use: [
      'file-loader',
    ],
  })

  config.module.rules.push({
    test: /\.(woff|woff2|eot|ttf|otf)$/,
    use: [
      'file-loader',
    ],
  })

  return config
}

```

**Comment by Sujimoshi** [comment]
I just managed to launch Storybook 4 with webpack 4 and react 15. As I mentioned above, I just set react and react-dom 16.4.1 as forward dependencies of `Storybook/core`, `Storybook/react` and `Storybook/ui`. So after the launch, there were two versions of the react inside the bundle. React 15 for my project, and React 16 for storybook components. And it works

**Comment by Sujimoshi** [comment]
Why react is listed as peerDependency, if Storybook generally use react for its ui. I'm surprised that this problem did not arise before

**Comment by ndelangen** [comment]
>  I just set react and react-dom 16.4.1 as forward dependencies of Storybook/core, Storybook/react and Storybook/ui. So after the start-storybook, there were two versions of the react inside the bundle, React 15 for my project, and React 16 for storybook components. And it works

Can you please help me understand what you mean with this?
I'm unfamiliar with "forward dependencies".

**Comment by Sujimoshi** [comment]
Direct dependency (just `dependencies` in `package.json`), sorry for my english

**Comment by ndelangen** [comment]
No problem, I think I understand now.

I guess if we set react & react-dom as regular dependencies with 16 or higher, this will fix this issue too.

Is what you're saying..?

---

Yeah that sounds like something we could do, to preserve React <16 compatibility!

@Hypnosphi is our peerDependency master, do you see any issues with this?

The version of react Storybook needs, is whatever emotion 10 needs.

**Comment by Sujimoshi** [comment]
Exactly

**Comment by Hypnosphi** [comment]
@Sujimoshi you have no plugins, so there are good chances that your setup will work with Storybook 3.4 and webpack 4 as is

**Comment by Hypnosphi** [comment]
We use React & React DOM in preview iframe as well: https://github.com/storybooks/storybook/blob/master/app/react/src/client/preview/render.js

**Comment by Sujimoshi** [comment]
> @norbert We use React & React DOM in preview iframe as well: https://github.com/storybooks/storybook/blob/master/app/react/src/client/preview/render.js

Is this a problem?

> We can actually use `resolve.alias` to enforce our version. But this means that components that use some APIs [removed in 16](https://github.com/facebook/react/blob/master/CHANGELOG.md#removed-deprecations) (like `React.PropTypes`) will break. Which can in turn be fixed by user-provided `resolve.alias` in custom webpack config =)

Not the best solution, as for me

> @Sujimoshi you have no plugins, so there are good chances that your setup will work with Storybook 3.4 and webpack 4 as is

Not an option, I have other configurations that require plug-ins

**Comment by Hypnosphi** [comment]
> I have other configurations that require plug-ins

those will require some duplication between `webpack.config.js` and `.storybook/webpack.config.js`

**Comment by Sujimoshi** [comment]
> > Is this a problem?
> 
> That means that we can render your components (created with React 15) using React DOM 16, which can become a problem in some cases

What if pass specific version of ReactDOM as option to storybook

**Comment by Hypnosphi** [comment]
I'd say that `resolve.alias` is such an option =)

**Comment by ndelangen** [comment]
As long as we keep the react in the iframe compatible with react15 we're good on react15 compatibility, IF we make sure iframe and manager can have a separate versions of react.

I think it's natural for there to be 2 versions of react there, otherwise react-users are locked to use the same version of react as we are.

**Comment by Hypnosphi** [comment]
I thinlk we can just prebuild manager and bundle our react 16 into dist

**Comment by Hypnosphi** [comment]
`addons.js` makes it a bit tricky but still possible

**Comment by Sujimoshi** [comment]
> > IF we make sure iframe and manager can have a separate versions of react.
> 
> A good question is, how do we do that. Does webpack allow different `require.resolve` for different entry points? Or we need two separate webpack builds?

Is I know webpack doesn't have such option.

**Comment by Hypnosphi** [comment]
And we probably can't use emotion in info addon unless we do #1501

**Comment by ndelangen** [comment]
@Hypnosphi docs-view will solve this. I'd say we're not doing #1501 and we'll assist people migrating to docs-view instead. (which will not have the react in preview)

**Comment by ndelangen** [comment]
> As I know webpack doesn't have such option. So you need two configs

Working that that exactly, will not make it into 4.0.0

But we'll likely introduce this in some 4.x.y

**Comment by shilman** [comment]
Released a first version of compatibility in 4.1.0-alpha.4: https://github.com/storybooks/storybook/releases/tag/v4.1.0-alpha.4

Still needs more testing and documentation.

**Comment by shilman** [comment]
cc @ranyitz @Sujimoshi @caitlin615 @arkadiusm @Buggytheclown @stereodenis 

If you'd like to test out the alpha on your projects and let us know if it works for you, that would be really awesome! Just try out `@storybook/react@next` which should get you the latest alpha.

@ndelangen is there anything else they need to do or will it work out of the box?

**Comment by ndelangen** [comment]
We changed so the preview and manager can use different versions.
With the last next release, using any version of react should just work.

**Comment by shilman** [comment]
@Frankcarvajal it's released as alpha in `@storybook/react@next`. Please check it out! We expect to do a stable release in the next few weeks.

**Comment by shlomitc** [comment]
@shilman - I've tried the new version on our project `wix-style-react` but building the storybook failsas Webpack fails to bundle the manager and hits undefined (no error checking on `build-static.js`).
Trying to log the webpack errors shows it couldn't resolve `react` from `react-dom`

You can check out the PR branch that was created https://github.com/wix/wix-style-react/pull/2494 and our discussion.

**Comment by ranyitz** [comment]
Thanks @shlomitc!

@shilman I've found the bug with the resolution of `react`/`react-dom`.

**TL;DR** The manager gets the user version of `react` and `react-dom` instead of the version specified in `@storybook/react`. The issue explained in details [in this comment](https://github.com/wix/wix-style-react/pull/2494#issuecomment-442661590).

**Comment by ndelangen** [comment]
The next "next" release should fix this.

**Comment by ndelangen** [comment]
@shilman Do we close this, as it's fixed now in "next"?

**Comment by shilman** [comment]
Yep, closing!

**Comment by shlomitc** [comment]
I keep getting errors with the latest alpha build (v4.1.0-alpha.11) when trying to run `build-stroybook`

```bash
ERR! Module not found: Error: Can't resolve 'react/lib/ReactCurrentOwner' in '/Users/shlomitc/WebstormProjects/wsr-storybook-new/node_modules/react-dom/lib'
(node:73176) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): ModuleNotFoundError: Module not found: Error: Can't resolve 'react/lib/ReactCurrentOwner' in '/Users/shlomitc/WebstormProjects/wsr-storybook-new/node_modules/react-dom/lib'
(node:73176) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by shilman** [comment]
Ta-da!! I just released https://github.com/storybooks/storybook/releases/tag/v4.1.6 containing PR #5148 that references this issue. Upgrade today to try it out!

**Comment by ndelangen** [comment]
Does storybook run when ignoring these warnings?

**Comment by alasdairhurst** [comment]
Yeah, it seems to. Not sure if it will run into any incompatibility when using additional features though

On 19 Jan 2019 01:52, Norbert de Langen <notifications@github.com> wrote:

Does storybook run when ignoring these warnings?

**Comment by ndelangen** [comment]
V5 should resolve these peerDependencies warnings
